enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,271 @@
+For that, my application container needs access to both shared and exclusive CPUs.
+
+  
+### Goals 
+* Provide a special-purpose at a runtime-level solution for containers in a pod to have the ability
+to request and subsequently be allocated both exclusive and shared CPUs.
+* Enable the telco-specific use case with a self-contained change,
+    minimizing the changes to the platform
+
+### Non-Goals
+* Introducing a generic mechanism in the platform that does involve Kubelet and pod spec changes.
+
+## Proposal
+
+Container processes are restricted
+to given [cpuset](https://man7.org/linux/man-pages/man7/cpuset.7.html) by their corresponding cgroup definition.
+Essentially, the kernel enforces the processes of container to run only at the CPU (core) ids specified under its
+cgroup.
+
+The mixed-cpu-node-plugin is a  runtime-agnostic, hybrid (composed of both [NRI-plugin](https://github.com/containerd/nri) and
+[device-plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)), 
+[mixed-cpu-node-plugin](https://github.com/openshift-kni/mixed-cpu-node-plugin) 
+that communicates with the container-runtime
+at various lifecycle stages of container e.g. CreateContainer, UpdateContainer, etc.,
+and modifies the container's cgroup to include a set of shared CPUs.
+
+The shared CPUs are part of the [reserved-cpus](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list) CPUs pool.
+The reason to utilize the reserved CPUs is because of CPU manager which is not aware of the CPUs
+lying in this pool, so it doesn't undo or changes the allocation logic performed by the plugin.
+Therefore, the plugin assigns the shared CPUs to containers without racing/conflict
+the CPU manager behavior. 
+More about this decision at the [Alternative](mixed-cpu-node-plugin.md#alternatives) section 
+
+The node-plugin populates a special device named `openshift.io/shared-cpus` to provide way for pods to request
+for this special type of CPUs.
+There's no meaning to the value/number of `openshift.io/shared-cpus` devices that the pod requests.
+For example:
+```yaml
+requests:
+   openshift.io/shared-cpus: 1
+```
+Or
+```yaml
+requests:
+   openshift.io/shared-cpus: 2
+```
+Are the same in that context, the container has access to the same number of shared-cpus in both cases.
+IOW, the device request only uses as a hint to tell the plugin that shared-cpus required for the container and
+not indicate any actual value.
+Once a container under the pod requests and gets access to shared cpus, new environment variable named
+`OPENSHIFT_SHARED_CPUS=<CPU-IDs>` will be present under the container's environment variables.
+
+The mixed-cpu-node-plugin itself is essentially a pod that runs as a DaemonSet on the worker nodes.
+cluster-node-tuning-operator (NTO) will be modified to support the deployment of the DaemonSet and additional
```

Comment:
So there will be 2 DaemonSets running in NTO namespace?

Reply:
If the feature is enabled, then yes. 
1. TuneD
2. Mixed-cpu-node-plugin (optional)

Reply:
I don't think there is a list of pods. The pod needs to be annotated with:
target.workload.openshift.io/management: {"effect":"PreferredDuringScheduling"}
See https://github.com/openshift/enhancements/blob/master/enhancements/workload-partitioning/management-workload-partitioning.md for details.

Reply:
OK but if WP is not enabled on the cluster?
Is the annotation should be depended on the cluster configuration?

Reply:
I think the pods are annotated regardless. If WP is not enabled, then the annotation is just ignored.